libindy/tests/crypto.rs

```diff
@@ -591,3 +592,100 @@ mod high_cases {
         }
     }
 }
+
+mod load {
+    extern crate rand;
+
+    use super::*;
+
+    use self::rand::{Rng, OsRng};
+
+    use std::cmp::max;
+    use std::thread;
+    use std::time::{Duration, SystemTime};
+
+    use utils::sequence::SequenceUtils;
+
+    const AGENT_CNT: usize = 10;
+    const DATA_SZ: usize = 10 * 1024;
+    const OPERATIONS_CNT: usize = 10;
+
+    /**
+     Environment varibales can be used for tuning this test:
+     - AGENTS_CNT - count of parallel agents
+     - OPERATIONS_CNT - operations per agent (consequence in same agent)
+     - DATA_SZ - data size for encryption
+     - UNENCRYPTED_WALLET - is wallet unencrypted (unencrypted by default)
+    */
+    #[test]
+    fn parallel_auth_encrypt() {
+        TestUtils::cleanup_storage();
+
+        let agent_cnt = std::env::var("AGENTS_CNT").ok().and_then(|s| s.parse::<usize>().ok()).unwrap_or(AGENT_CNT);
+        let data_sz = std::env::var("DATA_SZ").ok().and_then(|s| s.parse::<usize>().ok()).unwrap_or(DATA_SZ);
+        let operations_cnt = std::env::var("OPERATIONS_CNT").ok().and_then(|s| s.parse::<usize>().ok()).unwrap_or(OPERATIONS_CNT);
+        let credentials = if std::env::var("UNENCRYPTED_WALLET").is_ok() { None } else { Some(r#"{ "key": "test_passwd" }"#) };
+
+        let mut agents = Vec::new();
+        let mut os_rng = OsRng::new().unwrap();
+        for _ in 0..agent_cnt {
+            let wallet_name = format!("load-wallet-name-{}", SequenceUtils::get_next_id());
+            WalletUtils::create_wallet(POOL, &wallet_name, None, None, credentials).unwrap();
+            let wallet = WalletUtils::open_wallet(&wallet_name, None, credentials).unwrap();
+            let (_did, verkey) = DidUtils::create_and_store_my_did(wallet, None).unwrap();
+            let mut data = vec![0u8; data_sz];
+            os_rng.fill_bytes(&mut data.as_mut_slice());
+            agents.push((wallet, verkey, data));
+        }
+
+        let start_time = SystemTime::now();
+
+        let mut results = Vec::new();
+
+        for (wallet, verkey, data) in agents {
+            let thread = thread::spawn(move || {
+                let mut time_diffs = Vec::new();
+                for _ in 0..operations_cnt {
+                    let time = SystemTime::now();
+                    let _encrypted = CryptoUtils::auth_crypt(wallet, &verkey, &verkey, data.as_slice()).unwrap();
+                    let time_diff = SystemTime::now().duration_since(time).unwrap();
+                    time_diffs.push(time_diff);
+                }
+
+                WalletUtils::close_wallet(wallet).unwrap();
+                time_diffs
+            });
+            results.push(thread);
+        }
+
+        let mut all_diffs = Vec::new();
+        for result in results {
+            all_diffs.push(result.join().unwrap());
+        }
+        let total_duration = SystemTime::now().duration_since(start_time).unwrap();
```

Comment:
Total duraton should be calculated right after threads finish (after the `for` loop)

Reply:
@nveskovic I don't understand your comment, please add more details... It's calculated as you mentioned, right after `for` with joining to children threads. I see only one thing to optimize: unwrapping results and storing into Vec, but it's not significant, I think.

Reply:
Since @vimmerru already merged it **before** any additional changes were made, then there is not much use of a comment here. Nevertheless, what I was thinking is

Total duration should be calculated right after 
```
for (wallet, verkey, data) in agents {
...
}
```
not after additional
```
let mut all_diffs = Vec::new();
for result in results {
     all_diffs.push(result.join().unwrap());
}
```

Not a big difference but why not having accurate data if it is only a matter of sequence what's done first

hope it is clear now

Reply:
@nveskovic 

```for (wallet, verkey, data) in agents``` just finishes immediately as each loop iteration just starts new thread.  Next loop performs waiting for all threads and it is the correct place to put total time measurement.

Reply:
@nveskovic in other words, we should wait finishing all threads, so `.join()` call is required. Please see https://doc.rust-lang.org/std/thread/ .